sysboundary/ionosphere.cpp

```diff
@@ -1699,6 +1711,620 @@ namespace SBC {
+      std::vector<uint64_t> cellBWStepCounter(globalDccrgSize, 0);                                 /*!< Count number of field line tracing steps */
+      std::vector<int> cellNeedsContinuedFWTracing(globalDccrgSize, 1);                    /*!< Flag, whether tracing needs to continue on another task */
+      std::vector<int> cellNeedsContinuedBWTracing(globalDccrgSize, 1);                    /*!< Flag, whether tracing needs to continue on another task */
+      std::vector<std::array<Real, 3>> cellFWTracingCoordinates(globalDccrgSize);          /*!< In-flight node upmapping coordinates (for global reduction) */
+      std::vector<std::array<Real, 3>> cellBWTracingCoordinates(globalDccrgSize);          /*!< In-flight node upmapping coordinates (for global reduction) */
+      
+      // These guys are needed in the reductions at the bottom of the tracing loop.
+      std::vector<int> sumCellNeedsContinuedFWTracing(globalDccrgSize, 0);
+      std::vector<int> sumCellNeedsContinuedBWTracing(globalDccrgSize, 0);
+      std::vector<std::array<Real, 3>> sumCellFWTracingCoordinates(globalDccrgSize);
+      std::vector<std::array<Real, 3>> sumCellBWTracingCoordinates(globalDccrgSize);
+      std::vector<uint64_t> maxCellFWStepCounter(globalDccrgSize, 0);
+      std::vector<uint64_t> maxCellBWStepCounter(globalDccrgSize, 0);
+      
+      phiprof::start("first-loop");
+      for(int n=0; n<globalDccrgSize; n++) {
+         const CellID id = allDccrgCells[n];
+         cellFWTracingCoordinates.at(n) = mpiGrid.get_center(id);
+         cellBWTracingCoordinates.at(n) = cellFWTracingCoordinates.at(n);
+         
+         if(mpiGrid.is_local(id)) {
+            if(mpiGrid[id]->sysBoundaryFlag != sysboundarytype::NOT_SYSBOUNDARY) {
+               cellNeedsContinuedFWTracing[n] = 0;
+               cellNeedsContinuedBWTracing[n] = 0;
+               cellFWTracingCoordinates[n] = {0,0,0};
+               cellBWTracingCoordinates[n] = {0,0,0};
+            }
+         }
+      }
+      phiprof::stop("first-loop");
+      std::vector<int> minCellNeedsContinuedFWTracing(globalDccrgSize, 0);
+      std::vector<int> minCellNeedsContinuedBWTracing(globalDccrgSize, 0);
+      MPI_Allreduce(cellNeedsContinuedFWTracing.data(), minCellNeedsContinuedFWTracing.data(), globalDccrgSize, MPI_INT, MPI_MIN, MPI_COMM_WORLD);
+      MPI_Allreduce(cellNeedsContinuedBWTracing.data(), minCellNeedsContinuedBWTracing.data(), globalDccrgSize, MPI_INT, MPI_MIN, MPI_COMM_WORLD);
+      cellNeedsContinuedFWTracing = minCellNeedsContinuedFWTracing;
+      cellNeedsContinuedBWTracing = minCellNeedsContinuedBWTracing;
+      bool anyCellNeedsTracing;
+      
+      // Fieldline tracing function
+      TracingFieldFunction tracingFullField = [this, &perBGrid, &dPerBGrid, &technicalGrid](std::array<Real,3>& r, bool alongB, std::array<Real,3>& b)->void {
```

Comment:
There's plenty of code reuse and redundancy in this (now third) implementation of a TracingFieldFunction. Can this be somehow reduced?

Reply:
I am keeping this in the back of my head. I will have a thorough look again when I move to separate source files for the tracing shebang. However given that each of the four functions has different stopping conditions, and the fullbox guy does things double, forward and backward, and each of them has more or less distinct field interpolation lambdas, I suspect that reducing the redundancy of the base algorithm and passing a stack of different cases won't make it more readable.